src/python/pants/goal/run_tracker.py

```diff
@@ -390,3 +395,38 @@ def shutdown_worker_pool(self):
     N.B. This exists only for internal use and to afford for fork()-safe operation in pantsd.
     """
     SubprocPool.shutdown(self._aborted)
+
+  def report_target_info(self, scope, target, key, val):
+    """
+    Add target information to run_info under target_data
+    :param string scope: The running scope
+    :param string target: The target that we want to store info for
+    :param string or list of items to make into a key key: The key for the info being stored
+    :param dict or string val: The value of the info being stored
+    """
+    # Stringify list of keys to make one key
+    if not isinstance(key, basestring):
+      new_key = [str(key_item) for key_item in key]
+      key = '#'.join(new_key)
+
+    if target in self._target_data:
```

Comment:
A somewhat less error prone pattern for "create if not exists" is to do:
```
target_data = self._target_data.get(target, None)
if target_data is None:
  self._target_data[target] = target_data = {}
target_data[scope] ... # etc
```

Also possible to use `getdefault` or `DefaultDict`... `DefaultDict` might be the best option here.